### PR TITLE
Upon regaining focus, sync only if the bView has changed #457

### LINKED ISSUE #457: Upon regaining focus, sync only if the bView has changed
This should reduce the annoyance of syncing everytime HT gains focus.
One simple minded way to o detect if the bView has changed: HT can keep a copy of the HTML loaded to bView and compare it with the current HTML.

### PR DESCRIPTION
**Pull request by rd1992**
Will only sync when bview url changes
Fixes #457

### CODE REVIEW

**Comment by damithc** [comment]
Checking the URL may not be enough. For example, adding a label doesn't change the URL. Compare the full text instead?

**Comment by rd1992** [comment]
switching to comparing html source of the page. will leave some test comments

**Comment by rd1992** [comment]
test comment

**Comment by rd1992** [comment]
test comment

**Comment by rd1992** [comment]
test comment

**Comment by rd1992** [comment]
another test comment

**Comment by rd1992** [comment]
Has been fixed

File: src/ui/UI.java [hunk 1]
```diff
@@ -69,6 +69,9 @@
 	private MenuControl menuBar;
 	private BrowserComponent browserComponent;
 	private RepositorySelector repoSelector;
+	
+	// Bview changes
+	private String currentPageSource = "";
```
Comment (damithc) [inline src/ui/UI.java]: better name: `pageContentAtIssueLoading` ?  Otherwise line 188 looks odd.

Comment (rd1992) [inline src/ui/UI.java]: alright. changing to a better name.

**Comment by rd1992** [comment]
test comment

File: src/ui/UI.java [hunk 2]
```diff
@@ -69,6 +69,9 @@
 	private MenuControl menuBar;
 	private BrowserComponent browserComponent;
 	private RepositorySelector repoSelector;
+	
+	// Bview changes
+	private String pageContentOnLoad = "";
```
Comment (dariusf) [inline src/ui/UI.java]: Could this be kept track of in the browser component instead? That way it's concerned with its own state. The UI would call `browserComponent.hasPageChanged()` or something.

Comment (rd1992) [inline src/ui/UI.java]: good idea. i have updated it. you can merge it.

**Comment by damithc** [comment]
Seems to work only if the focus switching is done via alt+tab, but not via clicking.
e.g.
1.Select issue in HT
2. click on the page loaded in bView
3. Click on HT
At this point there should not be a sync, but it does trigger a sync.

**Comment by damithc** [comment]
Note that bView updated automatically after a while (e.g. timing info such as '5 minutes ago'). This means when HT gains focus after a long absence of focus, it immediately goes into a sync because the current bView doesn't match the previous snapshot. To counter that, we might have to refine the 'sync trigger check'. I'm not sure how though. On the other hand, this may not be a big problem for users who don't switch projects too often.

**Comment by rd1992** [comment]
@damithc: lets say you loaded the issue the first time and changed focus to bView. The check is only done when HT main window comes back to focus. So HT doesn't know yet that the url has changed in the bview. However any subsequent focus changes (for the same url) wont trigger a sync.

**Comment by rd1992** [comment]
as per your second comment. we sync every minute. Also if HT comes to focus after a long time i feel it should trigger a sync. (use case: Lets say i add an issue while your HT window was not in focus. So to see the issue a sync is required).

**Comment by rd1992** [comment]
the first issue you raised can be solved pretty easily. i probably misunderstood the requirement and implemented it this way.

**Comment by damithc** [comment]
Yes it is good to sync at that point. The problem is that project switching is disabled during sync. Many a time I click on HT on the task bar to bring it to front with the intention of switching projects, but I have to wait till the sync finishes. We disabled project switching during syncs as a workaround to a data corruption issue. It was a dirty fix which is causing grief in other ways. At some point we need to go back to the root cause (data corruption if project changes during a sync) and fix it.

**Comment by damithc** [comment]
The fact of the matter is syncs are good and we should do them as frequently as possible. It's just that we should do it without hampering other things.

**Comment by rd1992** [comment]
alright i have pushed the latest code. it does fix the issue you described

**Comment by dariusf** [comment]
The thing about the current design is that it assumes everything happens synchronously -- it wasn't a problem before we had caching as there was no state to be corrupted. We'll need to rewrite those parts (and design them better, probably with atomic updates, work queues for downloading, writing to file, etc.) to solve the problem at its root.

**Pull request merged**